src/node_file.cc

```diff
@@ -912,18 +916,18 @@ static void InternalModuleReadJSON(const FunctionCallbackInfo<Value>& args) {
       if (0 == memcmp(s, "exports", 7)) break;
     }
   }
-
-  Local<String> return_value;
-  if (p < pe) {
-    return_value =
-        String::NewFromUtf8(isolate,
-                            &chars[start],
-                            v8::NewStringType::kNormal,
-                            size).ToLocalChecked();
-  } else {
-    return_value = env->empty_object_string();
-  }
-
+  return_value->Set(
+    isolate->GetCurrentContext(),
+    FIXED_ONE_BYTE_STRING(isolate, "string"),
+    String::NewFromUtf8(isolate,
+                    &chars[start],
+                    v8::NewStringType::kNormal,
+                    size).ToLocalChecked()).Check();
+
+  return_value->Set(
+    isolate->GetCurrentContext(),
+    FIXED_ONE_BYTE_STRING(isolate, "containsKeys"),
+    Boolean::New(isolate, p < pe ? true : false)).Check();
```

Comment:
If the code here is kept-as is:

The `FIXED_ONE_BYTE_STRING()`s should be listed in the table in `env.h` instead and used as e.g. `env->contains_keys_string()`, because this code runs more than a handful of times during a Node.js process and that way we can avoid creating the strings over and over again.

Also, instead of `.Check()`s here we should be doing proper error handling, i.e. `if (return_value->Set(...).IsNothing()) return;`.

---

All that being said: The `return_values->Set()` calls will have a non-trivial performance impact (I don’t know how relevant that is here, but given that this function was introduced as an optimization, I assume that it is). A nicer alternative might be to return an `Array` from this C++ binding function with 2 entries, e.g.:

```c++
  Local<Value> return_value[] = {
    String::NewFromUtf8(...).ToLocalChecked(),
    Boolean::New(...)
  }
  args.GetReturnValue().Set(
      Array::New(isolate, return_value, arraysize(return_value)));
```

The downside is that this would not read quite as nicely from JS, but that could be addressed by destructuring in the wrapper module, I think:

```js
  const [ string, containsKeys ] = internalModuleReadJSON(path);
  cache.set(path, { string, containsKeys });
```

I think the rest of the code should keep working as it does in that case.